### ISSUE TITLE
Verbosity of Lockfile v6 causes PR viewer to truncate it

### ISSUE THREAD

**Issue by dmichon-msft** [pnpm/pnpm#7570]
### Verify latest release

- [ ] I verified that the issue exists in the latest pnpm release

### pnpm version

8.12.1

### Which area(s) of pnpm are affected? (leave empty if unsure)

Lockfile

### Reproduction steps

Have a large repository with several hundred projects and several thousand dependencies, including a significant number of external packages with a lot of peer dependencies.

Lockfile in v6 format is 5.7 MB.

### Describe the Bug

The verbosity of the v6 format in comparison to the v5 format results in a substantially longer lockfile, which causes PR viewing tools (GitHub, Azure DevOps) to truncate the file when viewing online, making lockfile-related issues difficult to spot. It also causes the appearance of erroneous diffs because an addition or removal will result in the truncation happening in a different point in the baseline vs. changed file.

### Expected Behavior

Lockfile should be less than 3.5 MB. It might be necessary to add a feature that splits the lockfile into separate files at deterministic (e.g. user-configured) lexicographic thresholds, or possibly just split `importers` and `packages` into separate physical files.

It could also help to use shorter identifiers than `specifier` and `version`, since these tokens are present tens of thousands of times.

### Which Node.js version are you using?

18.19.0

### Which operating systems have you used?

- [ ] macOS
- [ ] Windows
- [X] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

Ubuntu

**Comment by RexSkz** [comment]
Dividing the lock file into separate files does not solve the problem:

- According to the size limit: it will be more conflicts in lock files than we expected. The rest of the chunks can also be changed if there are new lines or removed lines in the first few chunks.
- Split the `importers` and `packages`: it may solve the 5.7MB file issue for now, but a larger monorepo can still have this issue.

Shortening the `specifier` and `version` can make them not readable and hard to debug. Currently, we can figure out the peers a package uses, we can know whether it's from a registry or a git repo. If we shorten them (maybe using hash?) we can't debug easily.

**Comment by zkochan** [comment]
What if we move some of the info out from the lockfile and store it outside of the repository. For instance, we could store all the package integrity checksums somewhere else.

**Comment by RexSkz** [comment]
Checksums can be moved to another file, but the file should still be in the project. Assume two users working on the same project: user A adds a package, commits and pushes; user B pulls and performs `pnpm i` and gets the wrong content (e.g. due to the MITM attack); without a correct checksum, there's no chance for B to get noticed.

**Comment by dmichon-msft** [comment]
Fair enough; my concern was mostly that the formatting change spiked the lockfile size significantly (on pnpm 7 it is small enough to view the diff; in pnpm 8 it truncates halfway through).

As-is I actually want more information in the lockfile, namely the effective specifiers for each (external) package's dependencies, not just what they resolved to, since that information is necessary for lockfile stability.

**Comment by zkochan** [comment]
Related: https://github.com/pnpm/pnpm/issues/7685

**Comment by zkochan** [comment]
This no longer reproduces on current pnpm, so closing.

The lockfile format has changed twice since, and version 9 is far smaller: entries are grouped and each package is listed once.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
